types/index.d.ts

```diff
@@ -0,0 +1,139 @@
+// Type definitions for @slack/events-api 1.0
+// Project: https://github.com/@slack/events-api
+// Definitions by: Tate Thurston <https://github.com/tatethurston>
+// Definitions: <https://github.com/@slack/events-api>
+
+import * as express from 'express';
+
+export function createSlackEventAdapter(verificationToken: string, options?: SlackEventsApi.Options): SlackEventsApi.Adapter;
+
+export namespace SlackEventsApi {
+  interface Options {
+    includeBody?: boolean;
+    includeHeaders?: boolean;
+    waitForResponse?: boolean;
+  }
+
+  interface Adapter {
+    expressMiddleware: (opts?: ExpressMiddlewareOptions) => ExpressMiddleware;
+    start: (port: string | number) => Promise<undefined>;
+    stop: () => Promise<string>;
+    on: OnEvent & OnError;
+  }
+
+  type ExpressMiddleware = (req: express.Request, res: express.Response, next: express.NextFunction) => void;
+
+  interface ExpressMiddlewareOptions {
+    propagateErrors?: boolean;
+  }
+
+  interface Event {
+    type: EventType;
+    event_ts: string;
+    // TODO: More constrained shape for the events
+    // There is conflicting information between
+    // https://api.slack.com/events-api#event_type_structure and individual
+    // events.
+    user?: string;
+    ts?: string;
+    item?: any;
+    [key: string]: any;
+  }
+
+  interface EventError {
+    code: ErrorCode;
+    message: string;
+    body: object;
+  }
+
+  type OnEvent = (event: EventType, Handler: EventHandler) => void;
+  type OnError = (event: "error", Handler: ErrorHandler) => void;
+
+  type EventHandler = (
+    event: Event,
+    body?: SlackEvent,
+    // TODO: More constrained types for headers and respond
+    headers?: any,
+    respond?: any
+  ) => void;
+
+  type ErrorHandler = (error: EventError) => void;
+
+  interface SlackEvent {
+    token: string;
+    team_id: string;
+    api_app_id: string;
+    event: Event;
+    type: "event_callback" | "url_verification";
+    authed_users: string[];
+    event_id: string;
+    event_time: number;
+  }
+
+  type ErrorCode =
+    "SLACKEVENTMIDDLEWARE_NO_BODY_PARSER" |
+    "SLACKEVENTMIDDLEWARE_TOKEN_VERIFICATION_FAILURE";
+
+  type EventType =
+    "app_uninstalled"         |
```

Comment:
i have some apprehension to committing an exhaustive list of event types into this repo. these values are bound to change over time, and that means developers would have to deal with a compiler error if an up to date set is not available. we came up with a technique to allow flexibility using generics [in this issue](https://github.com/slackapi/node-slack-sdk/issues/496). i'd prefer a solution that allowed unknown event types to work without a compiler error.

Reply:
Sure, if that's a goal then we could union String at the end or simply define EventType as a string.